src/util/ringdelaybuffer.h

```diff
@@ -0,0 +1,147 @@
+* ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+* CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/*
+* The text above constitutes the entire PortAudio license; however,
+* the PortAudio community also makes the following non-binding requests:
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#pragma once
+
+// The RingDelayBuffer code is based on the pa_ringbuffer.c
+// from the Portable Audio I/O Library. The template code is used
+// in the Mixxx.
+//
+// For the pa_ringbuffer.c see: mixxx/lib/portaudio/pa_ringbuffer.c
+//
+// Added functions:
+// - isFull
+// - isEmpty
+// - length
+// - getReadAvailable
+// - getWriteAvailable
+// - moveReadPositionBy
+//
+// Modified functions:
+// - clear
+// - read
+// - write
+
+#include "util/samplebuffer.h"
+#include "util/types.h"
+
+// The RingDelayBuffer is designed to be single-threaded
+// and based on that IS NOT THREAD SAVE.
+
+// TODO(davidchocholaty) Done documentation.
+
+/// The ring delay buffer allows moving with the reading position subject
+/// to certain rules. Another difference between the classic ring buffer is,
+/// that the ring delay buffer offers to read zero values
+/// which were not written by using the write method and write position.
+/// Both of these two specific properties are based on the cross-fading
+/// between changes of two delays.
+class RingDelayBuffer final {
+  public:
+    RingDelayBuffer(SINT bufferSize);
+
+    bool isFull() const {
+        return getWriteAvailable() == 0;
+    }
+
+    bool isEmpty() const {
+        return getReadAvailable() == 0;
+    }
+
+    void clear() {
+        m_fullFlag = false;
+        m_firstInputBuffer = true;
+
+        m_readPos = 0;
+        m_writePos = 0;
+
+        m_buffer.fill(0);
+    }
+
+    SINT size() const {
+        return m_buffer.size();
+    }
+
+    SINT getReadAvailable() const {
+        if (m_writePos > m_readPos) {
+            return (m_writePos - m_readPos);
+        } else if (m_writePos < m_readPos) {
+            return (m_buffer.size() - m_readPos) + m_writePos;
+        } else {
+            // The write position equals the read position (m_writePos == m_readPos).
+            // The buffer is full or empty (m_fullFlag).
+            if (m_fullFlag) {
+                return m_buffer.size();
+            }
+
+            return 0;
+        }
+    }
+
+    SINT getWriteAvailable() const {
+        return m_buffer.size() - getReadAvailable();
+    }
+
+    SINT read(CSAMPLE* pBuffer, const SINT numItems);
+    SINT write(const CSAMPLE* pBuffer, const SINT numItems);
+    SINT moveReadPositionBy(const SINT jumpSize);
```

Comment:
I don't understand why we should require the user code to do book keeping about the read position. The read position is IMHO an implementation detail that should be hidden. The user code has the desired delay already.

Reply:
I think I see what you are saying. In the specific context, this does make sense, but we have an important tradeoff to make here then. If we shift the delay book-keeping into the ringbuffer, the ringbuffer will loose its generality. The only advantage gained by this is possibly nicer code in the consumer, while sacrificing simplicity of the ringbuffer API. IMO there is a good chance the consumer will have to keep track of the latency anyways. And in that case, the only extra code required by the consumer is to call `moveReadPositionBy(oldDelay - newDelay)`.

Reply:
I would like to join this discussion. As @daschuer mentioned I agree that this code version is a little bit harder to use and the inner implementation is not so hidden. On the other side, as @Swiftb0y mentioned, my first goal was to make the API as general as possible. Anyway, if I'm thinking about it again, I can't think of any other specific use at the moment. If you can think of one, please share. IMO the RingDelayBuffer should be optimized for the purpose it was created first. It can be used as a normal ring buffer too, but I think, that for this purpose it should be better to use the [CircularBuffer](https://github.com/mixxxdj/mixxx/blob/070a475cdddf67b2423c564e3f7cc29ee8d8d455/src/util/circularbuffer.h). The `CircularBuffer` can be extended and improved as well of course for the "classic" purpose and I can take this task. But to my main thought which I would like to mention, the previous delay value for the purposed new version has to be stored in the `RingDelayBuffer` to avoid running complicated structure of reading position checks for every read run (as was mentioned in the discussion the check has to be in user code for the current version or in RingDelayBuffer code for the purposed version, so this check is the same). However, the `EngineEffectsDelay`, for which is RingDelayBuffer actually implemented, stores the previous delay value, so for the current use case the previous delay value information will be duplicated.

Reply:
Calling `moveReadPositionBy(oldDelay - newDelay)` is probably not what you want, because you need to read a buffer with the old delay first and than with the new delay for crossfading. So we require a `moveReadPositionBy(oldDelay - newDelay - numItems)`. The order of calls is also important. This is IMHO not understandable, without looking into the implementation. You need to understand what the read position is, and where it is initial. Otherwise the boundary checks may hit. 

The class is `RingDelayBuffer` and the API has nothing with delay in it, I am sure the first time user will expect a function with a delay parameter. Please keep also in mind that we need two consecutive reads, in case the delay has changed. 

Thinking about the proposed function again, It turns out that it will even release the implementation from keeping read position and a delay value as member. It allows to freely call API functions without order restrictions: 

```
SINT read(CSAMPLE* pBuffer, const SINT numItems, const SINT delayItems) {
    SINT readPos = m_writePos - numItems - delayItems;
    if (readPos < 0) {
       readPos + m_buffer.size(); 
    }
    ....
}
```

Reply:
Sure lets give the delay optimized version a try.

Reply:
I agree, sounds reasonable and good. It will be a lot cleaner approach. Thank you @daschuer.